Solution for Conway's Game of Life

fgs-pwright opened the pull request:
This solution includes a game that will be initialized with a square-tile grid and the standard Game of Life rules ("B2/S23").  Different rules could be set for the game, if desired.

Different implementations of the `IGrid` interface could be added to allow the game to be played on different types of surfaces, like a flat hexagonal grid, or even a polyhedron like a soccer ball.

fgs-pwright commented:
I'd like to generalize the game a little more by making the use of the `LifeState` enum generic in a few places, but I ran into some snags when I tried.  I may try to get some help with that.

FGS-BMarkham commented on GameOfLife/GameOfLife/SquareTileGrid.cs:
```diff
@@ -0,0 +1,72 @@
+            if (_rowDimension.IsOwnNeighbor())
+            {
+                yield return centerCellPosition;
+                yield return centerCellPosition;
```
I would expect `GetNeighborPositions` to return a _distinct_ set of `CellPosition`, for which this would surprise me. For cases of `IsOwnNeighbor`, I would suggest yielding the `centerCellPosition` a maximum of 1 times.

FGS-BMarkham commented on GameOfLife/GameOfLife/SquareTileGrid.cs:
```diff
@@ -0,0 +1,72 @@
+            }
+        }
+
+        private IEnumerable<CellPosition> GetPositionsExcluding(
```
Challenge:
LINQ has an `Exclude<T>(…)` method, which relies on equality. If you implement equality for `CellPosition`, you could probably simplify this down to a string of library calls that doesn't involve you doing your own loops.

FGS-BMarkham commented on GameOfLife/GameOfLife/ConsoleWriter.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace GameOfLife
+{
+    public class ConsoleWriter : IConsoleWriter
```
Good eye, introducing a seam here. Please note that this implementation doesn't have to use the `Console` directly. `System.Console.Out` and `System.Console.In` both return text stream manipulation abstractions that are super important in .NET, and you could just as well make your `IConsoleWriter` implementation take both of those.

Also, if you are going to have a `ReadLine`, better rename that interface to either `IConsole` or `IConsoleReaderWriter`.

FGS-BMarkham commented on GameOfLife/GameOfLife/Game.cs:
```diff
@@ -1,6 +1,116 @@
+        private List<LifeState[,]> _generations;
+
+        public const string GRID_NOT_LOADED = "A grid must be loaded before any steps can be taken.";
+        public const string INCOMPLETE_RULES = "Cannot step until rules have been defined for all possible cell states.";
```
Why are these public?

FGS-BMarkham commented on GameOfLife/GameOfLife/Game.cs:
```diff
@@ -1,6 +1,116 @@
+namespace GameOfLife
 {
-    public class Game
+    public class Game : IGame
```
This class has too many responsibilities. It's interacting with the user, it's managing engine initialization & execution. Please try to find a way to tear out these responsibilities so this class can be higher-level.

fgs-pwright replied:
Oh, I missed this comment the first time through.  I can see what you're saying, and I'll revise my solution to correct this.

FGS-BMarkham commented on GameOfLife/GameOfLife/SquareTileGrid.cs:
```diff
@@ -15,13 +16,29 @@ public SquareTileGrid(Cell[,] cells, bool wrapsOnRows, bool wrapsOnColumns)
+                    consoleWriter.Write(Environment.NewLine);
+                    column = 0;
+                }
+            }
```
This does not fit the responsibility paradigm of the class it was added to. Please consider implementing this elsewhere.

fgs-pwright replied:
I didn't have this method in this class initially, but I liked the idea of this class being the single source of information about the arrangement of cells or states, which is why I stuck it in here.

I moved this logic to the `Write...` method of the `Game` class, but I was thinking of creating a separate class altogether that would handle this.  Do you think that would be a better way to deal with it?

FGS-BMarkham replied:
Separate class is best, methinks.

fgs-pwright commented on GameOfLife/GameOfLife/IGrid.cs:
```diff
@@ -0,0 +1,11 @@
+    {
+        Cell GetCellAt(CellPosition cellPosition);
+        IEnumerable<Cell> GetNeighborsOfCellAt(CellPosition centerCellPosition);
+        IEnumerator<CellPosition> GetEnumerator();
```
Thanks.  In order to make this work, will I need to create a new class that implements `IEnumerator`, like the example given [here](https://msdn.microsoft.com/en-us/library/system.collections.ienumerable.getenumerator(v=vs.110).aspx#Anchor_2)?

FGS-BMarkham replied:
Good question! To answer it: Nope. I'm pretty sure that if you use the `yield` syntax, your method signature can either return `IEnumerable<T>` or `IEnumerator<T>`, and the compiler does the magic for you.

fgs-pwright replied:
Thanks again for this point.  It took me some forum scouring to put it together on my first try, and I'm not surprised I didn't do it exactly the best way.  How does it look now?

fgs-pwright commented on GameOfLife/GameOfLife/GridFactory.cs:
```diff
@@ -1,18 +1,30 @@
     public class GridFactory
     {
-        public Grid Create(LifeState[,] seed, params bool[] inputWrappingRules)
+        public SquareTileGrid CreateSquareTileGrid(LifeState[,] seed)
```
This is true.  My thought is that this would be a simple, working case that didn't require some of the complex validation you mentioned, and that the class could be extended with new `Create` methods, if that would be useful.

I'm going back and forth between thinking this is laziness on my part, or just thinking that this is a straightforward implementation that I can build on in the future.  What do you think?

FGS-BMarkham replied:
In the land of "user picks which kind of simulation to run", and options include:
- Non-cartesian systems
- Infinite planes (not representable by an array type)
- Variations in dimensions
… this code won't work. That said, none of those are in your requirements right now. Like I mentioned the other day - this is something you might want to come back to during CoP time.

fgs-pwright replied:
Good cases to consider.  I will almost certainly do that.

fgs-pwright commented on GameOfLife/GameOfLife/Rule.cs:
```diff
@@ -1,4 +1,6 @@
-﻿using System.Linq;
+﻿using System.Collections.Generic;
+using System.Linq;
+using static GameOfLife.LifeState;
```
The `using static ...` syntax is new?  I've mostly only used it for my enums so far, and I don't think I've been completely consistent with it.  I'll keep in mind that it's new as I'm using it in the future.

FGS-BMarkham replied:
For the first 14 years or so of C#, if you wanted to use another type's static members (enums too), you had to spell out the type names by hand. Each time. It sounds tedious, but some of us are used to _reading_ code like that. Highlights weird dependencies well. Like I said, I'll get over it. :-)

fgs-pwright replied:
That makes sense.  If you see me abusing this to obscure unhealthy dependencies, then let me know.  I'll keep that it mind when I use something like this in the future.

fgs-pwright commented on GameOfLife/GameOfLife/LifeState.cs:
```diff
@@ -2,7 +2,7 @@
-        Alive,
-        Dead
+        Dead,
+        Alive
```
I did this so that `Dead` would be the default value, for a number of reasons.  I'm happy to share some of them, if that would be a fruitful discussion.

FGS-BMarkham replied:
If you have engine rules or initialization behavior that indirectly relies on one of these values in a specific way, it is best to capture that in a named variable close to where it is being used. For example, mayhaps `GridFactory` should have a `DefaultValue` member `const` that points to `LifeState.Dead`. This will keep your code resilient against changes to that enum in the future.

FGS-BMarkham commented on GameOfLife/GameOfLife/IConsole.cs:
```diff
@@ -1,7 +1,11 @@
+    public interface IConsole
     {
+        TextReader In { get; set; }
+        TextWriter Out { get; set; }
```
You wouldn't put this on your interface, because it is an implementation detail, ain't it?

FGS-BMarkham commented on GameOfLife/GameOfLife/IGame.cs:
```diff
@@ -9,6 +9,6 @@ public interface IGame
         void Step();
         void WriteCurrentPatternToConsole();
-        IEnumerable<LifeState> GetCurrentPattern();
+        IEnumerable<IEnumerable<LifeState>> GetCurrentPattern();
```
When I saw this, it reminded me of a series of blog posts I wanted to share with the team: https://foxguard.slack.com/archives/dev-learning-emporium/p1485984405000172

Unless you are doing something really weird here (on purpose), you probably don't need an `IEnumerable<IEnumerable<T>>`. If you did this because you felt cornered into it by the structure of the implementation underneath, you should know that you can use the LINQ method `SelectMany` to flatten it down, or "unwrap" the monad, so-to-speak.

FGS-BMarkham commented on GameOfLife/GameOfLife/Game.cs:
```diff
@@ -0,0 +1,135 @@
+        private IConsole _console;
+        private IRuleset _rules;
+        private IGrid _grid;
+        private List<IEnumerable<IEnumerable<LifeState>>> _generations;
```
I'm not sure what this is supposed to be, but it kinda looks like you are saving a complete history of the run of the simulation. This won't work out well under continuous (or long-term) execution, and should probably be avoided.

The `IEnumerable<IEnumerable<LifeState>>` inside the list looks like an attempt to represent the state of a single generation of the grid without revealing the concrete types. Once again, this kinda reveals the jagged array implementation, and seems to be begging for a different kind of abstraction. (For example, what if you had an `IGameState` interface, which represented an immutable snapshot of gamestate? It would probably borrow elements from elsewhere in your application, so it would require some shuffling.)

FGS-BMarkham commented on GameOfLife/GameOfLife/Game.cs:
```diff
@@ -0,0 +1,135 @@
+
+namespace GameOfLife
+{
+    public class Game : IGame
```
Initializing, mutating, & validating a set of rules could probably exist in a class with a name more specific to that purpose. The word "Game" here might as well be as vague as "Program", so we need to justify its existence at every turn, and pull things out where we can.

FGS-BMarkham commented on GameOfLife/GameOfLifeConsole/Program.cs:
```diff
@@ -0,0 +1,15 @@
+        {
+            var consoleWriter = new TextReaderWriter();
+            var game = new Game(new RuleFactory(), new GridFactory(), consoleWriter, new Ruleset());
+            var consoleController = new ConsoleInterface(consoleWriter, game, new TextCommandParser(), new CommandRunner());
```
If only there existed some "thing" that could wire all these dependencies together for you, and inject them when you ask for them…
:-)
That can wait until Monopoly, but you'll want to do it early.

fgs-pwright replied:
Ha, yep.  Is there a DI framework that the team uses commonly, or one you'd suggest I check out?

fgs-dbudwin replied:
We tend to use Autofac.  We have used Ninject in the past though.

fgs-pwright replied:
Cool, thanks

fgs-dbudwin replied:
DI is weird at first...then you can't live without it.  Autofac is a bit of a more mature product than Ninject, but Ninject is slightly easier to work with.  You'll notice they have a lot of analogous actions between them.  I did my kata with Ninject but wished I had done it Autofac since that's what we actually use.  I would say it's only slightly more difficult to learn, especially for straight forward applications like katas.  YMMV.  We do still _technically_ use Ninject for 1 or 2 of our ~50 projects in PUMP.  But those projects are rarely touched which is why they haven't been switched over yet.  Let us know if you have any questions about this stuff.

FGS-BMarkham commented on GameOfLife/GameOfLifeTests/ConsoleInterfaceTests.cs:
```diff
@@ -0,0 +1,77 @@
+            VerifyDisplaysGrid();
+            VerifyPromptsUserForCommand();
+            VerifyParsesUserCommand();
+        }
```
Great use of decoupling and mocking to make sure your program behaves as expected!

FGS-BMarkham commented on GameOfLife/GameOfLifeTests/GameTests.cs:
```diff
@@ -0,0 +1,100 @@
+            var console = SetUpMockIConsole(textWriter);
+            var ruleset = SetUpMockIRuleset();
+
+            Game = new Game(ruleFactory, gridFactory, console, ruleset);
```
AutoFixture can create your test subjects, too. (Not just random values.) This is a really powerful thing that we use a lot in our projects.

We do this by leaning on two mechanisms.

The first mechanism we lean on is the fact that AutoFixture lets you `.Register(…)` things. When you do this, it will use that thing you registered anytime it encounters a constructor/property/etc that asks for that thing. This is really valuable if you want to wire up a mock that you've just spent pains configuring.

The second mechanism we lean on is the `AutoFixture.AutoMoq` NuGet package, (wire it up with a `Customize` call like you can see [here](http://blog.ploeh.dk/2010/08/19/AutoFixtureasanauto-mockingcontainer/)). This tells AutoFixture that every time it sees an interface or an abstract type in the constructor of a thing, if it doesn't know how to make one, to generate a mock for us.

This makes it really easy for our tests to mock what we need, ignore everything else, and not couple our tests to the layout of the constructor of the type under test.

Give it a poke!

fgs-dbudwin commented on GameOfLife/GameOfLifeConsole/Program.cs:
```diff
@@ -1,20 +1,36 @@
-                new GameAdvancer(consoleReaderWriter),
-                new GridWriter(consoleReaderWriter));
+            var builder = new ContainerBuilder();
+            builder.RegisterType<ConsoleReaderWriter>().As<IConsoleReaderWriter>().As<IConsoleWriter>();
```
This looks pretty on point with how we setup Autofac!

However, remember we like short `Main` methods that don't do much.  One suggestion is to take all of your registration out into a new file.  Our naming convention is `[Something]Module.cs`.  We often have many modules per project, but I think you can get away with one for simplicity sake.  Make the class inherit `Module` which is an Autofac class.  You'll be forced to re-implement a `Load` function where you can put all of your registrations.

fgs-pwright replied:
Good deal.  What's the significance of the "Something" in `[Something]Module.cs`?  I'm wondering what I should consider when I name this one.  Maybe something that captures the most high-level class registered, like `ConsoleGameModule.cs`?

fgs-dbudwin commented on GameOfLife/GameOfLifeConsole/Program.cs:
```diff
@@ -1,20 +1,36 @@
+            builder.RegisterType<GameAdvancer>().As<IGameAdvancer>();
+            builder.RegisterType<SquareTileGridWriter>().As<IGridWriter>();
+            builder.RegisterType<CommandRunner>().As<ICommandRunner>();
+            builder.RegisterType<TextCommandParser>().As<TextCommandParser>();
```
When you register a type against itself, you can use `AsSelf()`: ` builder.RegisterType<TextCommandParser>().AsSelf()`

fgs-dbudwin commented on GameOfLife/GameOfLifeConsole/Program.cs:
```diff
@@ -1,20 +1,36 @@
-                new Ruleset(new RuleFactory()),
-                new GridFactory(),
-                new GameAdvancer(consoleReaderWriter),
-                new GridWriter(consoleReaderWriter));
```
Where did all this injecting disappear to?  I don't see it in this commit...did it get moved into another commit?

fgs-dbudwin replied:
☝️ Ignore, haz dumbz

FGS-BMarkham commented:
Once upon a time, I gave feedback on a kata to the effect of: the Main method should be 4ish lines of code. Allow me to demonstrate it again:
```
using (var container = MySuperSpecialDependencyInjectionContainerFactory.Create())
{
    var mySpecialStartupObject = container.Resolve<IMySpecialStartupType>();
    mySpecialStartupObject.DoSomething();
}
```

Shoot for this.

fgs-pwright commented:
Ok, I moved my registration logic to a testable module, added a test class for it, and registered that module in a `Create` method of a static container factory class.

I wondered if I could test module registration separate from service resolution (ie. test that the factory `Create` method yielded a container with the module registered, and separately test that a container with that module registered resolved services as defined in the module), but it didn't look like a container would store any information about the _modules_ registered in it, but only the components.

> [Modules (...) are used to configure the container, they are not actually registered (...) like other components](http://autofac.readthedocs.io/en/latest/configuration/modules.html)

If there is a way I can or should test these separately, I'd be glad to do it but may need some hints.

FGS-BMarkham commented:
I have no further review comments for this PR.

However, this PR is unmergeable due to the target branch being `master`, and that not being the convention for our CodeCamp PRs. That makes :gopher: :squash: moot.

We'll just close the PR and pretend like we merged it. :-)
